onnxruntime/test/mlas/unittest/test_sqnbitgemm.cpp

```diff
@@ -0,0 +1,477 @@
+/*++
+
+Copyright (c) Microsoft Corporation. All rights reserved.
+
+Licensed under the MIT License.
+
+Module Name:
+
+    test_sqnbitgemm.h
+
+Abstract:
+
+    Tests for MLAS n-bit int block quantized GEMM.
+
+--*/
+
+#pragma once
+
+#include "test_util.h"
+#include "mlas_qnbit.h"
+
+namespace {
+
+constexpr size_t DivRoundUp(size_t a, size_t b) {
+  return (a + b - 1) / b;
+}
+
+constexpr size_t BlkDataSizeInBytes(size_t BlkBitWidth, size_t BlkLen) {
+  return BlkLen * BlkBitWidth / 8;
+}
+
+template <size_t BlkBitWidth>
+constexpr size_t ZeroPointsForBlksSizeInBytes(size_t BlkCount) {
+  if constexpr (BlkBitWidth <= 4) {
+    return DivRoundUp(BlkCount, 2);
+  } else {
+    return BlkCount;
+  }
+}
+
+template <size_t BlkBitWidth, size_t BlkLen>
+struct ReferenceQNBitPacking {
```

Comment:
note: this was mostly copied from here:
https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/contrib_ops/cpu/quantization/blockwise_quant_block.h

may be able to reuse the MLAS block quantization functions being added in #18101